Configure NIP-05 for nostr

satsie opened the pull request:
This PR adds the configuration needed for NIP-05. NIP-05 allows you to link a nostr public key with a domain name, basically creating a verified username that looks like an email address. This is our nostr account, currently with the handle bitcoindevs@iris.to: https://primal.net/p/npub10p33xu03t8q7d9nxtks63dq4gmt4v4d3ppd5lcdp4rg9hxsd0f8q7mn2l2

This is the guide I used to set this up, it's the same one I used for my personal account: https://nvk.org/n00b-nip5 There is one more step that I need to do after this is merged.

Emmanuel-Develops commented:
From the implementation documentation here: https://github.com/nostr-protocol/nips/blob/master/05.md

BDP has to serve the required file to whatever client queries for it.
- identifier format: `name@domain` e.g `bitcoindevs@bitcoindevs.xyz`
- constructed query: `https://{domain}/.well-known/nostr.json?name={name}` e.g `https://bitcoindevs.xyz/.well-known/nostr.json?name=bitcoindevs`

basically, we need a route for GET requests to that endpoint that returns the pubkey
we can have a static server (returns all pubkeys) or dynamic server (returns the pubkey for the name specified in the query parameters)

- To implement a static server, you can utilize the `/public` folder and have the JSON file in there
- To implement a dynamic server, configure a route.ts in `app/.well-known/nostr.json/route.ts` there you can pull the name from the search param (?name) and query where you store the pubkeys for the pubkey corresponding to the name provided
__note: `nostr.json` should be created as a folder__

Emmanuel-Develops reviewed (commented):
The guide is serving the files from github pages which is going to be a bit different here since we host on vercel and with next we can create routes.

This is the equivalent of dropping the json file in `/public/.well-known/nostr.json` (static server)

satsie commented:
Thank you so much @Emmanuel-Develops for the additional details on implementing NIP-05! Your explanations and instructions were very helpful. I made the following changes and rebased this PR:

- Update the pubkey to be in hex format instead of npub (good catch!)
- Update the name key to be "_", per the NIP: https://github.com/nostr-protocol/nips/blob/master/05.md#showing-just-the-domain-as-an-identifier
- Implement the dynamic server solution that returns the pubkey associated with the `name` query param. I did this in a new `route.ts` file and created a `route.test.ts` file to test the endpoint. I didn't see any other test frameworks in here so I installed jest. Let me know if this is okay. I only tested the success response and didn't add much error handling to the route code (such as if there is no `name` query param). For the route code I used what was in `app/subscribe/route.ts` as a guide.
- updates to `package.json` and the `bitcoin-topics` submodule came along for the ride

I also tested this locally with Postman:

![image](https://github.com/user-attachments/assets/65b37e5b-50e7-490c-a874-972d55f0885b)


I'm not very familiar with this code base, so let me know if there's anything I have overlooked. I'm glad you mentioned the static server option because it would be super cool if we could offer a `@bitcoindevs.xyz` identifier to members of the project :slightly_smiling_face:

Emmanuel-Develops commented:
You're welcome @satsie

I love the test, yes we do not have any testing built into BDP at the moment. I do fear jest is overkill for the API, but given its a devDependency it's fine (plus could encourage us to have more tests in the future)

> updates to package.json and the bitcoin-topics submodule came along for the ride

we do not update the sub-modules in PRs as it adds clutter and we have a [action](https://github.com/bitcoin-dev-project/bitcoin-dev-project/actions/workflows/trigger-deploy-from-submodule-update.yml) for updating it in prod when the sub-module changes. kindly force push without the public folder please.

Will review in a bit

Emmanuel-Develops commented on app/.well-known/nostr.json/route.ts:
```diff
@@ -0,0 +1,45 @@
+  try {
+    const name = request.nextUrl.searchParams.get('name') as string;
+
+    const file = await fs.readFile(process.cwd() + '/app/.well-known/nostr.json/nostr.json', 'utf8');
```
Tested against preview deployment at 
`https://bitcoin-dev-project-53brnx5ig-btc-knowledge-base.vercel.app/.well-known/nostr.json?name=_`

got 500 response:
```json
{
  "error": "ENOENT: no such file or directory, open '/var/task/app/.well-known/nostr.json/nostr.json'"
}
``` 
seems it has an error reading that file in prod.
You can try importing it instead.

i.e:
``` ts
import data from './nostr.json'
```

Emmanuel-Develops reviewed (changes requested):
Looks great!

Tested against preview deployment and got an error, more context in review comments

satsie commented on app/.well-known/nostr.json/route.ts:
```diff
@@ -0,0 +1,45 @@
+// See https://github.com/nostr-protocol/nips/blob/master/05.md#example
+export async function GET(request: NextRequest) {
+  try {
+    const name = request.nextUrl.searchParams.get('name') as string;
```
Thanks for providing the code! I decided to go with the latter, the 400 error. I know this code base is open source so returning all pubkeys is not a security risk, but it doesn't feel prudent to do so.

satsie commented:
Thank you for the very thorough and thoughtful review @Emmanuel-Develops! I made all the changes you mentioned but left the comments unresolved in case you want double check. To make it easy to review what changed, I made my updates a follow up commit instead of rebasing the original one.

I tested the following endpoints on prod and they behave as expected:
- https://bitcoin-dev-project-ptq6bcrh2-btc-knowledge-base.vercel.app/.well-known/nostr.json?name=_
- https://bitcoin-dev-project-ptq6bcrh2-btc-knowledge-base.vercel.app/.well-known/nostr.json?name=bitcoindevs
- https://bitcoin-dev-project-ptq6bcrh2-btc-knowledge-base.vercel.app/.well-known/nostr.json
- https://bitcoin-dev-project-ptq6bcrh2-btc-knowledge-base.vercel.app/.well-known/nostr.json?name=foo

Regarding the submodule update I committed by accident (and have since removed). I looked into how new contributors can avoid making the same mistake that I did and this update to my git config seemed to do the trick: `git config submodule.public/bitcoin-topics.ignore all`

I also think we can add `   ignore = dirty` to the `.gitmodules` file. Is that something you'd be open to? If so, just let me know and I can make a PR.

Emmanuel-Develops reviewed (changes requested):
Looks great!
tested `https://bitcoin-dev-project-ptq6bcrh2-btc-knowledge-base.vercel.app/.well-known/nostr.json?name=_`  and it works splendidly.

one more nit

Emmanuel-Develops commented:
That is a lot of testing, love it!

> update to my git config seemed to do the trick: git config submodule.public/bitcoin-topics.ignore all

interesting, but this applies to local git config yes?
I tried stop tracking sometime back but that deletes the submodule instead.

> I also think we can add ` ignore = dirty` to the `.gitmodules` file. Is that something you'd be open to? If so, just let me know and I can make a PR.

Okay, what does this fix?

satsie commented:
Thanks @Emmanuel-Develops! I addressed that nit. I want to do a bit more testing on how to make it easier for contributors to leave submodule updates alone. I will open a new PR or ticket with more detail on what the `ignore = dirty` setting does after I do more testing with it. Just in addressing the nit I accidentally pushed up submodule updates again :woman_facepalming: (I fixed it though)

Emmanuel-Develops reviewed (approved):
Looks good!
Merging to staging
